### PR TITLE
[Token/JWT] Add some leeway to jwt claims

### LINKED ISSUE #18880: Failed to push blob to local repo, error: http status code: 401
hi details about the setup, using 2.5.1 version on OpenShift 4.11, 

harbor-core:v2.5.1

So have configured proxy cache projects for upstream registries, when trying to pull images , we intermittently  see the below message and we dont see the repo created in the UI as well, the folders are visible when I check the volume. 

Unsuccessful 
_Failed to push blob to local repo, error: http status code: 401, body: {"errors":[{"code":"UNAUTHORIZED","message":"unauthorized to access repository:_

successful 
```
2023-06-28T09:51:51Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'InternalArtifact' on topic 'PUSH_ARTIFACT': ID-5, Repository-prod.registry.docker.site.gs.com/goharbor/harbor-exporter Tags-[v2.5.1] Digest-sha256:566be56fa88b377d858ec846832cef26542dfea1e5001b7986f9c22f3e6907b1 Operator-harbor#proxy-cache-service OccurAt-2023-06-28 09:51:51
2023-06-28T09:51:51Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'InternalArtifact' on topic 'PULL_ARTIFACT': ID-5, Repository-prod.registry.docker.site.gs.com/goharbor/harbor-exporter Tags-[v2.5.1] Digest-sha256:566be56fa88b377d858ec846832cef26542dfea1e5001b7986f9c22f3e6907b1 Operator- OccurAt-2023-06-28 09:51:51
2023-06-28T09:51:51Z [INFO] [/controller/event/handler/webhook/artifact/artifact.go:75]: []
2023-06-28T09:51:51Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'ArtifactWebhook' on topic 'PUSH_ARTIFACT': ID-5, Repository-prod.registry.docker.site.gs.com/goharbor/harbor-exporter Tags-[v2.5.1] Digest-sha256:566be56fa88b377d858ec846832cef26542dfea1e5001b7986f9c22f3e6907b1 Operator-harbor#proxy-cache-service OccurAt-2023-06-28 09:51:51
2023-06-28T09:51:51Z [INFO] [/controller/event/handler/webhook/artifact/artifact.go:75]: []
2023-06-28T09:51:51Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'ArtifactWebhook' on topic 'PULL_ARTIFACT': ID-5, Repository-prod.registry.docker.site.gs.com/goharbor/harbor-exporter Tags-[v2.5.1] Digest-sha256:566be56fa88b377d858ec846832cef26542dfea1e5001b7986f9c22f3e6907b1 Operator- OccurAt-2023-06-28 09:51:51
2023-06-28T09:51:51Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'AuditLog' on topic 'PULL_ARTIFACT': ID-5, Repository-prod.registry.docker.site.gs.com/goharbor/harbor-exporter Tags-[v2.5.1] Digest-sha256:566be56fa88b377d858ec846832cef26542dfea1e5001b7986f9c22f3e6907b1 Operator- OccurAt-2023-06-28 09:51:51
2023-06-28T09:51:51Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'Replication' on topic 'PUSH_ARTIFACT': ID-5, Repository-prod.registry.docker.site.gs.com/goharbor/harbor-exporter Tags-[v2.5.1] Digest-sha256:566be56fa88b377d858ec846832cef26542dfea1e5001b7986f9c22f3e6907b1 Operator-harbor#proxy-cache-service OccurAt-2023-06-28 09:51:51
2023-06-28T09:51:51Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'AuditLog' on topic 'PUSH_ARTIFACT': ID-5, Repository-prod.registry.docker.site.gs.com/goharbor/harbor-exporter Tags-[v2.5.1] Digest-sha256:566be56fa88b377d858ec846832cef26542dfea1e5001b7986f9c22f3e6907b1 Operator-harbor#proxy-cache-service OccurAt-2023-06-28 09:51:51
2023-06-28T09:51:51Z [INFO] [/pkg/notifier/notifier.go:205]: Handle notification with Handler 'P2PPreheat' on topic 'PUSH_ARTIFACT': ID-5, Repository-prod.registry.docker.site.gs.com/goharbor/harbor-exporter Tags-[v2.5.1] Digest-sha256:566be56fa88b377d858ec846832cef26542dfea1e5001b7986f9c22f3e6907b1 Operator-harbor#proxy-cache-service OccurAt-2023-06-28 09:51:51
```

### PR DESCRIPTION
**Pull request by an-toine**
# Comprehensive Summary of your change

On distributed systems, with Harbor deployed on multiple backend servers, it is not uncommon to observe some small discrepancies between server clocks, even with a reliable NTP source.

In such cases, a backend server may issue tokens with a `NotBefore` field set to a future timestamp for other servers, resulting in failed (403 unauthorized) image pulls for end users : 

```
[ERROR] [/pkg/token/token.go:68]: parse token error, token is not valid yet
```

As RFC7519 tolerates some leeway to be provided to account for clock skew, this PR is adding/substracting 10 seconds to fields `NotBefore`, `ExpiresAt` and `IssuedAt`.

With golang-jwt v5.2.0, a leeway field is directly integrated in the Validator struct (https://github.com/golang-jwt/jwt/blob/v5.2.0/validator.go#L38). This PR is just emulating this behavior for the release currently in use in Harbor.

# Issue being fixed
[Fixes #18880](https://github.com/goharbor/harbor/issues/18880)

Please indicate you've done the following:
- [X] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [X] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [X] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

**Review by Vad1mo — APPROVED** [review]

**Comment by Vad1mo** [comment]
Thank you for your contribution @an-toine, can you check if we could upgrade our dependency golang-jwt from v4 to v5?

**Comment by an-toine** [comment]
> Thank you for your contribution @an-toine, can you check if we could upgrade our dependency golang-jwt from v4 to v5?

I've opened https://github.com/goharbor/harbor/pull/19802 to migrate to golang-jwt v5.2.0 and use the built-in leeway attribute.

File: src/server/middleware/security/v2_token.go [hunk 1]
```diff
@@ -62,6 +65,11 @@ func (vt *v2Token) Generate(req *http.Request) security.Context {
 		return nil
 	}
 	cl := &v2TokenClaims{}
+	// Add some leeway to claims to account for clock skews on distributed environments
+	// https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.4
+	cl.Claims.RegisteredClaims.ExpiresAt = jwt.NewNumericDate(cl.Claims.RegisteredClaims.ExpiresAt.Time.Add(time.Second * 10))
```
Comment (MinerYang) [inline src/server/middleware/security/v2_token.go:70]: Hi @an-toine , 
we would like to keep this time frame as same as upstream distribution  which will be 60s.
https://github.com/distribution/distribution/blob/7c354a4b40feeea21d7eeae4de91c8ff7951e672/registry/auth/token/token.go#L25C4-L25C4
Could you help to update this and fix the CI issue ? Appreciate!

**Comment by MinerYang** [comment]
Hi @an-toine , 

We probably  do not need this PR if we add our leeway at the validate side. And we preferred to do this at the Validate side. Let's focus on review the another [PR](https://github.com/goharbor/harbor/pull/19802)

**Comment by an-toine** [comment]
> Hi @an-toine ,
> 
> We probably do not need this PR if we prefer to add our leeway at the validate side. Let's focus on review the another [PR](https://github.com/goharbor/harbor/pull/19802)

@MinerYang : fully agree, I will try to work on #19802 review by the end of the week.

**Pull request closed without merging**